ospd_openvas/daemon.py

```diff
@@ -247,6 +253,52 @@ def format_vt_modification_time(self, value: str) -> str:
 
         return datetime.utcfromtimestamp(int(value)).strftime("%Y%m%d%H%M%S")
 
+    def get_filtered_vts_list(self, vts, vt_filter: str) -> Optional[List[str]]:
+        """ Gets a collection of vulnerability test from the redis cache,
+        which match the filter.
+
+        Arguments:
+            vt_filter: Filter to apply to the vts collection.
+            vts: The complete vts collection.
+
+        Returns:
+            List with filtered vulnerability tests. The list can be empty.
+            None in case of filter parse failure.
+        """
+        if not vt_filter:
+            raise OspdCommandError('vt_filter: A valid filter is required.')
+
+        filters = self.parse_filters(vt_filter)
+        if not filters:
+            return None
+
+        if not self.nvti:
+            return None
+
+        vt_oid_list = [vtlist[1] for vtlist in self.nvti.get_oids()]
+        vt_oid_list_temp = copy.copy(vt_oid_list)
+        vthelper = VtHelper(self.nvti)
+
+        for _element, _oper, _filter_val in filters:
+            for vt_oid in vt_oid_list_temp:
+                if vt_oid not in vt_oid_list:
+                    continue
+
+                vt = vthelper.get_single_vt(vt_oid)
+                if vt is None or not vt.get(_element):
+                    vt_oid_list.remove(vt_oid)
+                    continue
+
+                _elem_val = vt.get(_element)
+                _val = self.format_filter_value(_element, _elem_val)
+
+                if self.filter_operator[_oper](_val, _filter_val):
+                    continue
+                else:
+                    vt_oid_list.remove(vt_oid)
+
+        return vt_oid_list
```

Comment:
I am really sure this code can be simplified by using the [filter function](https://docs.python.org/3/library/functions.html#filter). This would avoid having to copy the list.

Reply:
We need to do a copy because the original list will change during the iteration. Doing the original list smaller (removing the vts that do not match), speeds up the algorithm. And using the copy for the iteration, allows to use more than one filter over the already filtered list:
e.g. modification_time > and < than date
"<get_vts filter='modification_time&gt;202004140211;modification_time&lt;202004150211' details='1'/>"

Given the case we want to allow the use of only one filter, I tested with filter, removing the list copy and writing the valid vts in a new list. This takes longer (38 seconds with filter() against 17 second with the current algorithm).

The memory peak during the copy is ~6 MB for some seconds
According to my tests, implementing the filter fun

Reply:
Thanks for the investigation. Seems your solution is better then the filter one.